src/Utilities/ConstantExpressions.hpp

```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
 
 /// \ingroup ConstantExpressionsGroup
 namespace ConstantExpressions_detail {
-template <typename T, int N, typename = std::nullptr_t>
-struct pow;
 
-/// \cond HIDDEN_SYMBOLS
-template <typename T, int N>
-struct pow<T, N, Requires<(N > 0)>> {
-  SPECTRE_ALWAYS_INLINE static constexpr decltype(auto) apply(const T& t) {
-    return t * pow<T, N - 1>::apply(t);
-  }
-};
+// Implementation functions for the pow template function below which computes
+// optimized powers where the exponent is a compile-time integer
 
-template <typename T, int N>
-struct pow<T, N, Requires<(N < 0)>> {
-  SPECTRE_ALWAYS_INLINE static constexpr decltype(auto) apply(const T& t) {
-    return static_cast<T>(1) / (t * pow<T, -N - 1>::apply(t));
-  }
-};
+// base case power 0: Returns 1.0. This execution path will only be followed if
+// called by the recursive power assembling functions below. This will need to
+// be overloaded for types for which the simple return is inappropriate.
+template <typename T>
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& /*t*/, std::integral_constant<int, 0> /*meta*/,
+    cpp17::bool_constant<true> /*meta*/) noexcept {
+  return 1.0;
+}
+// duplicate required for both execution paths, and template can't be used to
+// allow for priority in overload resolution
+template <typename T>
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& /*t*/, std::integral_constant<int, 0> /*meta*/,
+    cpp17::bool_constant<false> /*meta*/) noexcept {
+  return 1.0;
+}
 
+// special case power 1: acts as a direct identity function for efficiency
 template <typename T>
-struct pow<T, 0, Requires<not blaze::IsVector<T>::value>> {
-  SPECTRE_ALWAYS_INLINE static constexpr T apply(const T& /*t*/) {
-    return static_cast<T>(1);
-  }
-};
-/// \endcond
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& t, std::integral_constant<int, 1> /*meta*/,
+    cpp17::bool_constant<false> /*meta*/) noexcept {
+  return t;
+}
+
+// general case for positive powers: return the power via recursive inline call,
+// which expands to a series of multiplication operations.
+template <int N, typename T>
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& t, std::integral_constant<int, N> /*meta*/,
+    cpp17::bool_constant<true> /*meta*/) noexcept {
+  return t * pow_impl(t, std::integral_constant<int, N - 1>{},
+                          cpp17::bool_constant<true>{});
+}
+
+// general case for negative powers: return the multiplicative inverse of the
+// result from the recursive inline call, which expands to a series of
+// multiplication operations. This assumes that division is supported with
+// tt::get_fundamental_type_t<T> and T. If not, this utility will need further
+// specialization.
+template <int N, typename T>
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& t, std::integral_constant<int, N> /*meta*/,
+    cpp17::bool_constant<false> /*meta*/) noexcept {
```

Comment:
... and this one should not do a multiplication and should just call the `-N` version for the denominator.  Finally, the `pow`  function itself should dispatch using `cpp17::bool_constant<(N >= 0)>`.

The reasons for the extra multiplications fell into three groups.  Because the `bool_constant` changed between 1 and 2, the `N >= 2` cases were using 0 instead of 1 as the base case, computing t^1 as `t * 1` instead of `t`.  The `N == -1` case was explicitly computing its denominator as `t * pow<0>(t)`.  The `N <= -3` cases were fine themselves, but called the `N >= 2` cases and so inherited their bug.

Reply:
I tried it and got the same result as you.  But it has nothing to do with the recursion (which should all be optimized away anyway).  If I just evaluate `r = a * b`, `r = (a * b) * 1.0`, and `r = a * (b * 1.0)` with `r`, `a`, and `b` all DataVectors of the same size (100 for my test), the two triple products are both *faster* than the double product.  All the expressions are faster if `b` is replaced by `a` to get a squaring expression, but calculating the square by `a * a` is still the slowest.

@nilsdeppe do you have any idea what is going on here?

Reply:
The new version's complexity is largely to avoid SFINAE and to grab some extra speed when computing `pow<1>`. If the speed difference for the *1.0 small enough and mysterious enough to ignore, I'll revert to the code that @wthrowe described, which has the intuitive optimizations minimizing number of multiplications.